### DISCUSSION TITLE
thread/cas

### DISCUSSION THREAD

**Discussion by giscus** [itwanger/tobebetterjavaer-giscus#235] [Announcements]
# thread/cas

Java实现CAS的原理

https://tobebetterjavaer.com/thread/cas.html

**Comment by SkMoran** [comment]
这里看不懂了 举点应用的例子更好一些

**Comment by LFAW233** [comment]
private volatile Pair<V> pair;

    /**
     * Creates a new {@code AtomicStampedReference} with the given
     * initial values.
     *
     * @param initialRef the initial reference
     * @param initialStamp the initial stamp
     */
    public AtomicStampedReference(V initialRef, int initialStamp) {
        pair = Pair.of(initialRef, initialStamp);
    }

    public boolean compareAndSet(V   expectedReference,
                                 V   newReference,
                                 int expectedStamp,
                                 int newStamp) {
        Pair<V> current = pair;
        return
            expectedReference == current.reference &&
            expectedStamp == current.stamp &&
            ((newReference == current.reference &&
              newStamp == current.stamp) ||
             casPair(current, Pair.of(newReference, newStamp)));
    }

补一下pair的申明逻辑。
此外这边真正试图解决ABA问题的只有引入的Stamp，其他的一长串判断实际上是用来减少CAS被执行的次数，从而减少开销提升性能。

**Comment by ExcuMe** [comment]
ABA问题既然使用了时间戳或者版本号，感觉旧值预期值的比较没有意义了，直接判断时间戳或者版本号不久可以分辨出其他线程有没有更新更新变量么，这点不是很理解。
